File: src/tools/enumpoly/enumpoly.cc
//
// This file is part of Gambit
// Copyright (c) 1994-2025, The Gambit Project (https://www.gambit-project.org)
//
// FILE: src/tools/enumpoly/enumpoly.cc
// Enumerates all Nash equilibria via support enumeration
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 2 of the License, or
// (at your option) any later version.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to the Free Software
// Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
//

#include <iostream>
#include <fstream>
#include <cstdlib>
#include <getopt.h>
#include "gambit.h"
#include "solvers/enumpoly/enumpoly.h"

using namespace Gambit;
using namespace Gambit::Nash;

int g_numDecimals = 6;
bool g_verbose = false;

void PrintBanner(std::ostream &p_stream)
{
  p_stream << "Compute Nash equilibria by solving polynomial systems\n";
  p_stream << "Gambit version " VERSION ", Copyright (C) 1994-2025, The Gambit Project\n";
  p_stream << "This is free software, distributed under the GNU GPL\n\n";
}

void PrintHelp(char *progname)
{
  PrintBanner(std::cerr);
  std::cerr << "Usage: " << progname << " [OPTIONS] [file]\n";
  std::cerr << "If file is not specified, attempts to read game from standard input.\n";
  std::cerr << "With no options, reports all Nash equilibria found.\n\n";

  std::cerr << "Options:\n";
  std::cerr << "  -d DECIMALS      show equilibrium probabilities with DECIMALS digits\n";
  std::cerr << "  -h, --help       print this help message\n";
  std::cerr << "  -S               use strategic game\n";
  std::cerr << "  -m MAXREGRET     maximum regret acceptable as a proportion of range of\n";
  std::cerr << "                   payoffs in the game\n";
  std::cerr << "  -e EQA           terminate after finding EQA equilibria\n";
  std::cerr << "                   (default is to search in all supports)\n";
  std::cerr << "  -q               quiet mode (suppresses banner)\n";
  std::cerr << "  -V, --verbose    verbose mode (shows supports investigated)\n";
  std::cerr << "  -v, --version    print version information\n";
  std::cerr << "                   (default is only to show equilibria)\n";
  exit(1);
}

void PrintProfile(std::ostream &p_stream, const std::string &p_label,
                  const MixedStrategyProfile<double> &p_profile)
{
  p_stream << p_label;
  for (size_t i = 1; i <= p_profile.MixedProfileLength(); i++) {
    p_stream.setf(std::ios::fixed);
    p_stream << ',' << std::setprecision(g_numDecimals) << p_profile[i];
  }
  p_stream << std::endl;
}

void PrintSupport(std::ostream &p_stream, const std::string &p_label,
                  const StrategySupportProfile &p_support)
{
  if (!g_verbose) {
    return;
  }
  p_stream << p_label;
  for (auto player : p_support.GetGame()->GetPlayers()) {
    p_stream << ",";
    for (auto strategy : player->GetStrategies()) {
      p_stream << ((p_support.Contains(strategy)) ? 1 : 0);
    }
  }
  p_stream << std::endl;
}

void PrintProfile(std::ostream &p_stream, const std::string &p_label,
                  const MixedBehaviorProfile<double> &p_profile)
{
  p_stream << p_label;
  for (size_t i = 1; i <= p_profile.BehaviorProfileLength(); i++) {
    p_stream.setf(std::ios::fixed);
    p_stream << "," << std::setprecision(g_numDecimals) << p_profile[i];
  }
  p_stream << std::endl;
}

void PrintSupport(std::ostream &p_stream, const std::string &p_label,
                  const BehaviorSupportProfile &p_support)
{
  if (!g_verbose) {
    return;
  }
  p_stream << p_label;
  for (auto player : p_support.GetGame()->GetPlayers()) {
    for (auto infoset : player->GetInfosets()) {
      p_stream << ",";
      for (auto action : infoset->GetActions()) {
        p_stream << ((p_support.Contains(action)) ? 1 : 0);
      }
    }
  }
  p_stream << std::endl;
}

int main(int argc, char *argv[])
{
  opterr = 0;

  bool quiet = false;
  bool useStrategic = false;
  double maxregret = 1.0e-4;
  int stopAfter = 0;

  int long_opt_index = 0;
  struct option long_options[] = {{"help", 0, nullptr, 'h'},
                                  {"version", 0, nullptr, 'v'},
                                  {"verbose", 0, nullptr, 'V'},
                                  {nullptr, 0, nullptr, 0}};
  int c;
  while ((c = getopt_long(argc, argv, "d:hSm:e:qvV", long_options, &long_opt_index)) != -1) {
    switch (c) {
    case 'v':
      PrintBanner(std::cerr);
      exit(1);
    case 'd':
      g_numDecimals = atoi(optarg);
      break;
    case 'h':
      PrintHelp(argv[0]);
      break;
    case 'S':
      useStrategic = true;
      break;
    case 'm':
      maxregret = atof(optarg);
      break;
    case 'e':
      stopAfter = atoi(optarg);
      break;
    case 'q':
      quiet = true;
      break;
    case 'V':
      g_verbose = true;
      break;
    case '?':
      if (isprint(optopt)) {
        std::cerr << argv[0] << ": Unknown option `-" << ((char)optopt) << "'.\n";
      }
      else {
        std::cerr << argv[0] << ": Unknown option character `\\x" << optopt << "`.\n";
      }
      return 1;
    default:
      abort();
    }
  }

  if (!quiet) {
    PrintBanner(std::cerr);
  }

  std::istream *input_stream = &std::cin;
  std::ifstream file_stream;
  if (optind < argc) {
    file_stream.open(argv[optind]);
    if (!file_stream.is_open()) {
      std::ostringstream error_message;
      error_message << argv[0] << ": " << argv[optind];
      perror(error_message.str().c_str());
      exit(1);
    }
    input_stream = &file_stream;
  }

  try {
    const Gambit::Game game = Gambit::ReadGame(*input_stream);
    if (!game->IsPerfectRecall()) {
      throw Gambit::UndefinedException(
          "Computing equilibria of games with imperfect recall is not supported.");
    }

    if (!game->IsTree() || useStrategic) {
      EnumPolyStrategySolve(
          game, stopAfter, maxregret,
          [](const MixedStrategyProfile<double> &eqm) { PrintProfile(std::cout, "NE", eqm); },
          [](const std::string &label, const StrategySupportProfile &support) {
            PrintSupport(std::cout, label, support);
          });
    }
    else {
      EnumPolyBehaviorSolve(
          game, stopAfter, maxregret,
          [](const MixedBehaviorProfile<double> &eqm) { PrintProfile(std::cout, "NE", eqm); },
          [](const std::string &label, const BehaviorSupportProfile &support) {
            PrintSupport(std::cout, label, support);
          });
    }
    return 0;
  }
  catch (std::runtime_error &e) {
    std::cerr << "Error: " << e.what() << std::endl;
    return 1;
  }
}

File: src/pygambit/cli/enumpoly.py
#
# This file is part of Gambit
# Copyright (c) 1994-2026, The Gambit Project (https://www.gambit-project.org)
#
# FILE: src/pygambit/cli/enumpoly.py
# Command-line driver program for Nash equilibrium enumeration via support
# enumeration and polynomial system solving
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA.
#
"""Command-line driver program for Nash equilibrium enumeration via support
enumeration and polynomial system solving.

.. note::

   Solving via PHCpack (the C++ tool's ``phcpack_path`` equivalent has no CLI
   option) is not available here; see :func:`pygambit.nash.enumpoly_solve`'s
   `phcpack_path` parameter for the equivalent computation from Python.
"""

from __future__ import annotations

import click

import pygambit as gbt

from .common import (
    handle_errors,
    load_game,
    render_profile_csv,
    render_support_csv,
    validate_stop_after,
    version_option,
)

DESCRIPTION = "Compute Nash equilibria by solving polynomial systems"
PROG_NAME = "gambit-enumpoly"


@click.command(
    context_settings={"help_option_names": ["-h", "--help"]},
    help=(
        f"{DESCRIPTION}.\n\n"
        "Reads a game from FILE, or from standard input if FILE is not specified. "
        "With no options, reports all Nash equilibria found."
    ),
)
@click.argument("file", required=False, default=None)
@click.option(
    "-d",
    "decimals",
    default=6,
    show_default=True,
    type=int,
    help="show equilibrium probabilities with DECIMALS digits",
)
@click.option("-S", "--strategic", is_flag=True, help="use strategic game")
@click.option(
    "-m",
    "maxregret",
    default=1.0e-8,
    show_default=True,
    type=float,
    help="maximum regret acceptable as a proportion of the range of payoffs in the game",
)
@click.option(
    "-e",
    "stop_after",
    type=str,
    default=None,
    callback=validate_stop_after,
    help="terminate after finding EQA equilibria (default is to search in all supports)",
)
@click.option(
    "-r",
    "max_rectangles",
    default=20_000,
    show_default=True,
    type=int,
    help="maximum number of rectangles to examine when searching for roots on a "
    "single support, before giving up on that support",
)
@click.option("-q", "--quiet", is_flag=True, help="quiet mode (suppresses banner)")
@click.option(
    "-V",
    "--verbose",
    is_flag=True,
    help="verbose mode (shows supports investigated); default is only to show equilibria",
)
@version_option(DESCRIPTION)
@handle_errors
def main(
    file: str | None,
    decimals: int,
    strategic: bool,
    maxregret: float,
    stop_after: int | None,
    max_rectangles: int,
    quiet: bool,
    verbose: bool,
) -> None:
    game = load_game(quiet, DESCRIPTION, file, PROG_NAME)
    if not game.is_perfect_recall:
        raise ValueError("Computing equilibria of games with imperfect recall is not supported.")

    def render(profile) -> None:
        click.echo(render_profile_csv(profile, "NE", decimals))

    def render_event(event) -> None:
        if not verbose:
            return
        if isinstance(event, gbt.EnumPolyCandidateSupportEvent):
            click.echo(render_support_csv(event.support, "candidate"))
        elif isinstance(event, gbt.EnumPolySingularSupportEvent):
            click.echo(render_support_csv(event.support, "singular"))
        elif isinstance(event, gbt.EnumPolyBudgetExceededSupportEvent):
            click.echo(render_support_csv(event.support, "budget-exceeded"))

    gbt.nash.enumpoly_solve(
        game,
        use_strategic=strategic,
        stop_after=stop_after,
        maxregret=maxregret,
        max_rectangles=max_rectangles,
        nash_callback=render,
        event_callback=render_event,
    )


if __name__ == "__main__":
    main()
